pkg/scheduler/plugins/nodereservation/plugin.go

```diff
@@ -0,0 +1,201 @@
+func (f *Plugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// computePodResourceRequest returns a framework.Resource that covers the largest
+// width in each resource dimension. Because init-containers run sequentially, we collect
+// the max in each dimension iteratively. In contrast, we sum the resource vectors for
+// regular containers since they run simultaneously.
+//
+// Example:
+//
+// Pod:
+//   InitContainers
+//     IC1:
+//       CPU: 2
+//       Memory: 1G
+//     IC2:
+//       CPU: 2
+//       Memory: 3G
+//   Containers
+//     C1:
+//       CPU: 2
+//       Memory: 1G
+//     C2:
+//       CPU: 1
+//       Memory: 1G
+//
+// Result: CPU: 3, Memory: 3G
+func computePodResourceRequest(pod *corev1.Pod) *preFilterState {
+	result := &preFilterState{}
+	for _, container := range pod.Spec.Containers {
+		result.Add(container.Resources.Requests)
+	}
+
+	// take max_resource(sum_pod, any_init_container)
+	for _, container := range pod.Spec.InitContainers {
+		result.SetMaxResource(container.Resources.Requests)
+	}
+
+	// If Overhead is being utilized, add to the total requests for the pod
+	if pod.Spec.Overhead != nil {
+		result.Add(pod.Spec.Overhead)
+	}
+
+	return result
+}
+
+func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
+		return nil, fmt.Errorf("error reading %q from cycleState: %w", preFilterStateKey, err)
+	}
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to NodeResourcesFit.preFilterState error", c)
+	}
+	return s, nil
+}
+
+func (f *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	podQos := apiext.GetPodQoSClass(pod)
+	cpuReservedByNode := *resource.NewMilliQuantity(0, resource.DecimalSI)
+	node := nodeInfo.Node()
+	if node != nil {
+		if reserved, ok := node.Annotations[apiext.ReservedByNode]; ok {
+			cpus, qos := util.GetCPUsReservedByNode(reserved)
+			if qos == podQos {
+				cpuReservedByNode = cpus
+			}
+		}
+	}
+
+	s, err := getPreFilterState(cycleState)
+	if err != nil {
+		return framework.AsStatus(err)
+	}
+
+	insufficientResources := fitsRequest(s, nodeInfo, cpuReservedByNode)
+
+	if len(insufficientResources) != 0 {
+		// We will keep all failure reasons.
+		failureReasons := make([]string, 0, len(insufficientResources))
+		for _, r := range insufficientResources {
+			failureReasons = append(failureReasons, r.Reason)
+		}
+		return framework.NewStatus(framework.Unschedulable, failureReasons...)
+	}
+
+	return nil
+}
+
+type InsufficientResource struct {
+	resschedplug.InsufficientResource
+	reservedByNode int64
+}
+
+func fitsRequest(podRequest *preFilterState, nodeInfo *framework.NodeInfo, cpuReservedByNode resource.Quantity) []InsufficientResource {
```

Comment:
the node reservation supports reserving multiple resources, so we can support them together.